In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('../data/data.csv')

/tmp/ipykernel_101668/785670083.py:1: DtypeWarning: Columns (62) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/data.csv')


In [3]:
df.shape

(148741, 64)

# Tratamento

### Lista de remoção por ter muitos valores nulos

In [4]:
# colunas com muitos nulos ruim para analise
descartados_nulos = [
    "outrosSintomas",
    "outrasCondicoes",
    "cbo",
    "condicoes",
    "codigoBuscaAtivaAssintomatico",
    "outroBuscaAtivaAssintomatico",
    "codigoTriagemPopulacaoEspecifica",
    "outroTriagemPopulacaoEspecifica",
    "outroLocalRealizacaoTestagem",
    "codigoEstadoTeste2",
    "codigoTipoTeste2",
    "codigoFabricanteTeste2",
    "source_id",
    "codigoResultadoTeste2",
    "codigoEstadoTeste3",
    "codigoTipoTeste3",
    "codigoFabricanteTeste3",
    "codigoResultadoTeste3",
    "codigoEstadoTeste4",
    "codigoTipoTeste4",
    "codigoFabricanteTeste4",
    "codigoResultadoTeste4",
    "dataColetaTeste2",
    "dataColetaTeste3",
    "dataColetaTeste4",
]

In [5]:
# Remover colunas com muitos nulos
df = df.drop(columns=descartados_nulos)

In [6]:
print(f"Shape após remoção: {df.shape}")

Shape após remoção: (148741, 39)


### removendo por redundancia

In [7]:
lista_reduntantes = [
    "municipioIBGE",
    "estadoIBGE",
    "estadoNotificacao",
    "municipioNotificacao",
    "municipioNotificacaoIBGE",
    "estadoNotificacaoIBGE",
    "municipioIBGE",
]

In [8]:
# Remover colunas redundantes
df = df.drop(columns=lista_reduntantes)

# Tratamento bools

In [9]:
# Converter "Sim"/"Não" para booleano (True/False)
df["profissionalSaude"] = df["profissionalSaude"].map({"Sim": True, "Não": False})
df["profissionalSeguranca"] = df["profissionalSeguranca"].map({"Sim": True, "Não": False})


In [10]:
# Converter 1=Sim, 2=Não para booleano
df["codigoContemComunidadeTradicional"] = df["codigoContemComunidadeTradicional"].map({1: True, 2: False})


In [11]:
# Converter codigoRecebeuVacina para booleano diretamente na coluna original
# 1=Sim=True, 2=Não=False, 3=Ignorado=None
mapa_vacina = {1: True, 2: False, 3: None}
df["codigoRecebeuVacina"] = df["codigoRecebeuVacina"].map(mapa_vacina)
print("codigoRecebeuVacina (convertido para booleano):")
print(df["codigoRecebeuVacina"].value_counts(dropna=False))

codigoRecebeuVacina (convertido para booleano):
codigoRecebeuVacina
True     96988
NaN      28085
False    23564
None       104
Name: count, dtype: int64


In [12]:
df.head(5)

sintomas profissionalSaude  racaCor  \
0                                  Assintomático             False  Amarela   
1                                  Assintomático             False    Parda   
2  Coriza, Tosse, Dor de Garganta, Dor de Cabeça             False    Parda   
3            Dor de Cabeça, Tosse, Febre, Outros             False    Parda   
4                           Tosse, Dor de Cabeça             False   Branca   

  profissionalSeguranca       sexo estado municipio       origem  \
0                 False   Feminino   Pará   Tucuruí  parse-cloud   
1                 False  Masculino   Pará   Tucuruí  parse-cloud   
2                 False   Feminino   Pará   Tucuruí  parse-cloud   
3                 False   Feminino   Pará   Tucuruí  parse-cloud   
4                 False   Feminino   Pará   Tucuruí  parse-cloud   

               evolucaoCaso                classificacaoFinal  ...  \
0                      Cura  Síndrome Gripal Não Especificada  ...   
1                      Cura  Síndrome Gripal Não Especificada  ...   
2  Em tratamento domiciliar           Confirmado Laboratorial  ...   
3                      Cura  Síndrome Gripal Não Especificada  ...   
4                      Cura  Síndrome Gripal Não Especificada  ...   

   dataInicioSintomas  dataEncerramento dataPrimeiraDose dataSegundaDose  \
0                 NaN        2024-04-11       2021-04-08      2021-07-08   
1                 NaN        2024-03-27       2021-06-21      2021-09-21   
2          2022-01-15        2024-05-29       2021-08-14      2021-09-14   
3          2022-07-10        2024-04-24       2021-08-17      2021-09-14   
4          2022-01-29        2024-05-28       2021-06-09      2021-09-09   

  codigoEstadoTeste1 codigoTipoTeste1 codigoFabricanteTeste1  \
0                3.0              3.0                  715.0   
1                3.0              3.0                  715.0   
2                3.0              3.0                  715.0   
3                3.0              3.0                  792.0   
4                3.0              3.0                  715.0   

  codigoResultadoTeste1 dataColetaTeste1 idade  
0                   2.0       09/07/2022   NaN  
1                   2.0       31/08/2022   NaN  
2                   1.0       18/01/2022  29.0  
3                   2.0       12/07/2022  24.0  
4                   2.0       30/01/2022  28.0  

[5 rows x 33 columns]

# Tratamento de campos categóricos (códigos)

In [13]:
print(df.isnull().sum())

sintomas                                 4
profissionalSaude                        4
racaCor                                108
profissionalSeguranca                  175
sexo                                     0
estado                                   0
municipio                                0
origem                                   0
evolucaoCaso                         23210
classificacaoFinal                   23929
codigoEstrategiaCovid                 7653
codigoLocalRealizacaoTestagem         7651
codigoRecebeuVacina                  28189
codigoLaboratorioPrimeiraDose        57369
codigoLaboratorioSegundaDose         68508
lotePrimeiraDose                     57371
loteSegundaDose                      68508
codigoContemComunidadeTradicional     5687
excluido                                 0
validado                                 0
codigoDosesVacina                    53570
totalTestesRealizados                    0
dataNotificacao                          8
dataInicioS

## Estratégia para tratamento de nulos

**Campos OBRIGATÓRIOS - Remover linhas se nulo:**
- `sexo`, `estado`, `municipio` - Dados demográficos básicos (0 nulos ✅)
- `dataNotificacao` - Data essencial (8 nulos - remover)
- `totalTestesRealizados` - Informação crítica (0 nulos ✅)
- `origem`, `excluido`, `validado` - Campos do sistema (0 nulos ✅)

**Campos CLÍNICOS - Manter nulos (informação ausente é válida):**
- `sintomas` (4), `racaCor` (108), `profissionalSaude` (4), `profissionalSeguranca` (175)
- `evolucaoCaso` (23210), `classificacaoFinal` (23929) - Casos em andamento
- `dataEncerramento` (24982) - Só preenche quando fecha o caso

**Campos de TESTAGEM - Manter nulos (nem todos fazem teste):**
- `codigoEstrategiaCovid` (7653), `codigoLocalRealizacaoTestagem` (7651)
- `codigoEstadoTeste1` (16720), `codigoTipoTeste1` (16720)
- `codigoFabricanteTeste1` (38068), `codigoResultadoTeste1` (20697)
- `dataColetaTeste1` (18831)

**Campos de VACINAÇÃO - Manter nulos (nem todos vacinados):**
- `codigoRecebeuVacina` (28085), `codigoDosesVacina` (53570)
- `dataPrimeiraDose` (57974), `dataSegundaDose` (69270)
- `codigoLaboratorioPrimeiraDose` (57369), `codigoLaboratorioSegundaDose` (68508)
- `lotePrimeiraDose` (57371), `loteSegundaDose` (68508)

**Campos PROBLEMÁTICOS:**
- `dataInicioSintomas` (60249) - **CRÍTICO!** Muitos nulos, mas é campo importante
- `idade` (39347) - **CRÍTICO!** Dado demográfico essencial
- `codigoContemComunidadeTradicional` (5687) - Aceitar nulo ou preencher como False?

In [14]:
# Tratamento de nulos (preparação para dashboard e banco de dados)

print(f"Total de registros ANTES: {len(df)}")

# 1. Remover linhas com campos obrigatórios nulos
df = df.dropna(subset=['dataNotificacao'])
print(f"Após remover dataNotificacao nula: {len(df)}")

# 2. Preencher booleanos nulos como False (para dashboard não quebrar)
df['codigoContemComunidadeTradicional'] = df['codigoContemComunidadeTradicional'].fillna(False).infer_objects(copy=False)
df['profissionalSaude'] = df['profissionalSaude'].fillna(False).infer_objects(copy=False)
df['profissionalSeguranca'] = df['profissionalSeguranca'].fillna(False).infer_objects(copy=False)
df['codigoRecebeuVacina'] = df['codigoRecebeuVacina'].fillna(False)

# 3. Categorias texto: "Não Informado" para visualização
df["racaCor"] = df["racaCor"].fillna("Ignorado")
df['evolucaoCaso'] = df['evolucaoCaso'].fillna('Em Andamento')
df['classificacaoFinal'] = df['classificacaoFinal'].fillna('Não Classificado')
df['sintomas'] = df['sintomas'].fillna('Assintomático')

# 4. Códigos categóricos: 0 = "Não Informado" (tabelas de domínio)
df['codigoEstrategiaCovid'] = df['codigoEstrategiaCovid'].fillna(0)
df['codigoLocalRealizacaoTestagem'] = df['codigoLocalRealizacaoTestagem'].fillna(0)
df['codigoDosesVacina'] = df['codigoDosesVacina'].fillna(0)
df['codigoFabricanteTeste1'] = df['codigoFabricanteTeste1'].fillna(0)
df['codigoResultadoTeste1'] = df['codigoResultadoTeste1'].fillna(0)
df['codigoEstadoTeste1'] = df['codigoEstadoTeste1'].fillna(0)
df['codigoTipoTeste1'] = df['codigoTipoTeste1'].fillna(0)
df['codigoLaboratorioPrimeiraDose'] = df['codigoLaboratorioPrimeiraDose'].fillna(0)
df['codigoLaboratorioSegundaDose'] = df['codigoLaboratorioSegundaDose'].fillna(0)

# 5. Lotes de vacina: "Não Informado"
df['lotePrimeiraDose'] = df['lotePrimeiraDose'].fillna('Não Informado')
df['loteSegundaDose'] = df['loteSegundaDose'].fillna('Não Informado')

# 6. Regra especial: dataInicioSintomas
# Se nula, usar dataNotificacao (casos assintomáticos)
df.loc[df['dataInicioSintomas'].isna(), 'dataInicioSintomas'] = df.loc[df['dataInicioSintomas'].isna(), 'dataNotificacao']

# 7. MANTER NULL: Datas opcionais (linha do tempo epidemiológica)
# - dataPrimeiraDose, dataSegundaDose, dataColetaTeste1, dataEncerramento
# - idade (dado demográfico ausente é válido)

print(f"\n✅ Tratamento epidemiológico concluído!")
print(f"Registros mantidos: {len(df)}")
print(f"\nNulos restantes (apenas datas/idade opcionais):")
print(df.isnull().sum()[df.isnull().sum() > 0])

Total de registros ANTES: 148741
Após remover dataNotificacao nula: 148733

✅ Tratamento epidemiológico concluído!
Registros mantidos: 148733

Nulos restantes (apenas datas/idade opcionais):

✅ Tratamento epidemiológico concluído!
Registros mantidos: 148733

Nulos restantes (apenas datas/idade opcionais):


/tmp/ipykernel_101668/2953931297.py:10: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['codigoContemComunidadeTradicional'] = df['codigoContemComunidadeTradicional'].fillna(False).infer_objects(copy=False)
/tmp/ipykernel_101668/2953931297.py:11: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['profissionalSaude'] = df['profissionalSaude'].fillna(False).infer_objects(copy=False)
/tmp/ipykernel_101668/2953931297.py:12: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result

dataEncerramento    24980
dataPrimeiraDose    57966
dataSegundaDose     69262
dataColetaTeste1    18829
idade               39345
dtype: int64


In [15]:
print(df.isnull().sum())

sintomas                                 0
profissionalSaude                        0
racaCor                                  0
profissionalSeguranca                    0
sexo                                     0
estado                                   0
municipio                                0
origem                                   0
evolucaoCaso                             0
classificacaoFinal                       0
codigoEstrategiaCovid                    0
codigoLocalRealizacaoTestagem            0
codigoRecebeuVacina                      0
codigoLaboratorioPrimeiraDose            0
codigoLaboratorioSegundaDose             0
lotePrimeiraDose                         0
loteSegundaDose                          0
codigoContemComunidadeTradicional        0
excluido                                 0
validado                                 0
codigoDosesVacina                        0
totalTestesRealizados                    0
dataNotificacao                          0
dataInicioS

# Tratamento de tipos 

In [16]:
print(df.dtypes)

sintomas                              object
profissionalSaude                       bool
racaCor                               object
profissionalSeguranca                   bool
sexo                                  object
estado                                object
municipio                             object
origem                                object
evolucaoCaso                          object
classificacaoFinal                    object
codigoEstrategiaCovid                float64
codigoLocalRealizacaoTestagem        float64
codigoRecebeuVacina                     bool
codigoLaboratorioPrimeiraDose         object
codigoLaboratorioSegundaDose          object
lotePrimeiraDose                      object
loteSegundaDose                       object
codigoContemComunidadeTradicional       bool
excluido                              object
validado                              object
codigoDosesVacina                     object
totalTestesRealizados                  int64
dataNotifi

In [17]:
# Converter colunas de datas para datetime, mantendo nulos
colunas_datas = [
    "dataNotificacao",
    "dataInicioSintomas",
    "dataEncerramento",
    "dataPrimeiraDose",
    "dataSegundaDose",
    "dataColetaTeste1",
]

for coluna in colunas_datas:
    df[coluna] = pd.to_datetime(df[coluna], errors="coerce")

print(df.dtypes)
print(f"Shape após conversão (mantendo nulos): {df.shape}")

sintomas                                     object
profissionalSaude                              bool
racaCor                                      object
profissionalSeguranca                          bool
sexo                                         object
estado                                       object
municipio                                    object
origem                                       object
evolucaoCaso                                 object
classificacaoFinal                           object
codigoEstrategiaCovid                       float64
codigoLocalRealizacaoTestagem               float64
codigoRecebeuVacina                            bool
codigoLaboratorioPrimeiraDose                object
codigoLaboratorioSegundaDose                 object
lotePrimeiraDose                             object
loteSegundaDose                              object
codigoContemComunidadeTradicional              bool
excluido                                     object
validado    

In [18]:
df["codigoRecebeuVacina"].value_counts(dropna=False)

codigoRecebeuVacina
True     96988
False    51745
Name: count, dtype: int64

In [19]:
df.to_csv('../data/main/data_tratado.csv', index=False)